node/src/store/sled_store.rs

```diff
@@ -0,0 +1,497 @@
+use std::convert::Infallible;
+use std::io;
+use std::ops::Deref;
+use std::path::Path;
+
+use async_trait::async_trait;
+use celestia_types::hash::Hash;
+use celestia_types::ExtendedHeader;
+use directories::ProjectDirs;
+use sled::transaction::{ConflictableTransactionError, TransactionError};
+use sled::{Db, Error as SledError, Transactional, Tree};
+use tempdir::TempDir;
+use tendermint_proto::Protobuf;
+use tracing::debug;
+
+use crate::store::Store;
+use crate::store::{Result, StoreError};
+
+const HEAD_HEIGHT_KEY: &[u8] = b"KEY.HEAD_HEIGHT";
+const HASH_TREE_ID: &[u8] = b"HASH";
+const HEIGHT_TO_HASH_TREE_ID: &[u8] = b"HEIGHT";
+
+#[derive(Debug)]
+pub struct SledStore {
+    db: Db,
+    headers: Tree,
+    height_to_hash: Tree,
+}
+
+impl SledStore {
+    pub fn new(network_id: &str) -> sled::Result<Self> {
+        let Some(project_dirs) = ProjectDirs::from("co", "eiger", "celestia") else {
+            return Err(sled::Error::Io(io::Error::new(
+                io::ErrorKind::Other,
+                "Failed to get cache directory",
+            )));
+        };
+        let mut db_path = project_dirs.cache_dir().to_owned();
+        db_path.push(network_id);
+
+        Self::new_in_path(db_path)
+    }
+
+    pub fn new_temp() -> sled::Result<Self> {
+        let tmp_path = TempDir::new("celestia")?.into_path();
+
+        let db = sled::Config::default()
+            .path(tmp_path)
+            .temporary(true)
+            .create_new(true) // make sure we fail if db is already there
+            .open()?;
+        Self::init(db)
+    }
+
+    pub fn new_in_path<P>(path: P) -> sled::Result<Self>
+    where
+        P: AsRef<Path>,
+    {
+        let db = sled::open(path)?;
+        Self::init(db)
+    }
+
+    fn init(db: Db) -> sled::Result<Self> {
+        let headers = db.open_tree(HASH_TREE_ID)?;
+        let height_to_hash = db.open_tree(HEIGHT_TO_HASH_TREE_ID)?;
+
+        Ok(Self {
+            db,
+            headers,
+            height_to_hash,
+        })
+    }
+
+    pub fn head_height(&self) -> Result<u64> {
+        read_height_by_db_key(&self.db, HEAD_HEIGHT_KEY)
+    }
+
+    pub fn get_by_hash(&self, hash: &Hash) -> Result<ExtendedHeader> {
+        read_header_by_db_key(&self.headers, hash.as_bytes())
+    }
+
+    pub fn get_by_height(&self, height: u64) -> Result<ExtendedHeader> {
+        let hash = read_hash_by_db_key(&self.height_to_hash, &height_to_key(height))?;
+
+        self.get_by_hash(&hash)
+    }
+
+    pub fn get_head(&self) -> Result<ExtendedHeader> {
+        let head_height = self.head_height()?;
+        self.get_by_height(head_height)
+    }
```

Comment:
Does sled do any caching? If not, then we need to cache lasted appended header and use it as head.